File: backend/intelligence/decision_engine.py
# backend/intelligence/decision_engine.py
# SwiftAid — AI decision engine
#
# Responsibilities:
#   1. Signal pre-emption  — clear N intersections ahead for each ambulance
#   2. Conflict resolution — when two units head for the same node,
#                            higher priority wins, lower waits briefly
#   3. Predictive reroute  — if upcoming congestion exceeds threshold,
#                            flag the ambulance for rerouting
#
# No ML here — pure rule-based logic that is easy to explain to judges.

from utils.constants import CONFIG


class DecisionEngine:
    """
    Central AI coordinator called every simulation tick.
    Works with SignalManager and CityGrid to make decisions.
    """

    def __init__(self, signal_manager, grid):
        self.signals            = signal_manager
        self.grid               = grid
        self.conflicts_resolved = 0

    def update(self, ambulances: list) -> dict | None:
        """
        Main update method — call once per tick.

        Returns:
            {'reroute': <Ambulance>}  if a unit needs rerouting
            None                      otherwise
        """
        # Step 1: pre-empt signals for each moving ambulance
        for amb in ambulances:
            if amb.state == 'moving':
                self._preempt_ahead(amb)

        # Step 2: resolve conflicts between units
        if len(ambulances) > 1:
            self._resolve_conflicts(ambulances)

        # Step 3: check if any unit needs rerouting due to congestion
        for amb in ambulances:
            if amb.state == 'moving' and self._should_reroute(amb):
                return {'reroute': amb}

        return None

    # ── Signal pre-emption ───────────────────────────────────────────
    def _preempt_ahead(self, amb):
        """
        Pre-clear the next PREEMPT_LOOKAHEAD intersections.
        Skips edge nodes (row 0, row G, col 0, col G) — those have no signals.
        """
        lookahead = CONFIG['PREEMPT_LOOKAHEAD']
        for look in range(1, lookahead + 1):
            ni = amb.path_idx + look
            if ni >= len(amb.path):
                break
            r, c = amb.path[ni]
            if 1 <= r < self.grid.G and 1 <= c < self.grid.G:
                new_override = self.signals.preempt(r, c, amb.id)
                if new_override:
                    amb.sigs_overridden += 1

    # ── Conflict resolution ──────────────────────────────────────────
    def _resolve_conflicts(self, ambulances: list):
        """
        Compare every pair of ambulances.
        If their upcoming paths overlap, higher priority unit gets signal
        clearance; lower priority unit waits briefly.
        Simple pairwise check — fine for ≤3 units.
        """
        ahead = CONFIG['PREEMPT_LOOKAHEAD'] + 1
        for i in range(len(ambulances)):
            for j in range(i + 1, len(ambulances)):
                a = ambulances[i]
                b = ambulances[j]
                if a.state != 'moving' or b.state != 'moving':
                    continue

                # Upcoming nodes for each unit
                a_ahead = set(tuple(a.path[k]) for k in range(a.path_idx, min(a.path_idx + ahead, len(a.path))))
                b_ahead = set(tuple(b.path[k]) for k in range(b.path_idx, min(b.path_idx + ahead, len(b.path))))

                if a_ahead & b_ahead:   # paths overlap
                    winner = a if a.priority <= b.priority else b
                    loser  = b if winner is a else a
                    # Give winner's next intersection green
                    if winner.path_idx < len(winner.path):
                        r, c = winner.path[winner.path_idx]
                        self.signals.preempt(r, c, winner.id)
                    # Loser waits
                    loser.waiting    = True
                    loser.wait_timer = 30
                    self.conflicts_resolved += 1

    # ── Predictive reroute check ─────────────────────────────────────
    def _should_reroute(self, amb) -> bool:
        """
        Look ahead 3 edges. If average congestion exceeds REROUTE_THRESHOLD,
        signal that this ambulance needs a new path.
        This simulates lightweight prediction without any ML model.
        """
        avg = self.grid.average_congestion_ahead(amb.path, amb.path_idx, horizon=3)
        return avg > CONFIG['REROUTE_THRESHOLD']

File: frontend/static/js/decision_engine.js
// intelligence/decision_engine.js — Signal priority + conflict resolution logic

class DecisionEngine {
  constructor(signalManager, grid) {
    this.signals = signalManager;
    this.grid = grid;
    this.conflictsResolved = 0;
  }

  // Main per-tick update for a list of ambulances
  update(ambulances) {
    const preemptEnabled = document.getElementById('tog-preempt')?.checked ?? true;
    const rerouteEnabled = document.getElementById('tog-reroute')?.checked ?? true;
    const conflictEnabled = document.getElementById('tog-conflict')?.checked ?? true;

    for (const amb of ambulances) {
      if (amb.state !== 'moving') continue;

      // 1. Signal pre-emption: clear next PREEMPT_LOOKAHEAD signals
      if (preemptEnabled) {
        for (let look = 1; look <= CONFIG.PREEMPT_LOOKAHEAD; look++) {
          const idx = amb.pathIdx + look;
          if (idx >= amb.path.length) break;
          const { r, c } = amb.path[idx];
          if (r >= 1 && r < this.grid.G && c >= 1 && c < this.grid.G) {
            if (this.signals.preempt(r, c, amb.id)) {
              amb.sigsOverridden++;
            }
          }
        }
      }
    }

    // 2. Conflict detection: check if two ambs heading to same intersection
    if (conflictEnabled && ambulances.length > 1) {
      for (let i = 0; i < ambulances.length; i++) {
        for (let j = i + 1; j < ambulances.length; j++) {
          const a = ambulances[i], b = ambulances[j];
          if (a.state !== 'moving' || b.state !== 'moving') continue;
          if (this._pathsConflict(a, b)) {
            this.conflictsResolved++;
            this.signals.resolveConflict(a.r, a.c, a, b);
          }
        }
      }
    }

    // 3. Dynamic reroute: check if current path became heavily congested
    if (rerouteEnabled) {
      for (const amb of ambulances) {
        if (amb.state !== 'moving') continue;
        if (this._shouldReroute(amb)) {
          return { reroute: amb }; // caller handles rerouting
        }
      }
    }

    return null;
  }

  _pathsConflict(a, b) {
    const ahead = CONFIG.PREEMPT_LOOKAHEAD + 1;
    const aNodes = a.path.slice(a.pathIdx, a.pathIdx + ahead).map(n => `${n.r},${n.c}`);
    const bNodes = b.path.slice(b.pathIdx, b.pathIdx + ahead).map(n => `${n.r},${n.c}`);
    return aNodes.some(n => bNodes.includes(n));
  }

  _shouldReroute(amb) {
    // Check next 3 nodes — if avg congestion > threshold, reroute
    const horizon = 3;
    let total = 0, count = 0;
    for (let i = amb.pathIdx; i < Math.min(amb.pathIdx + horizon, amb.path.length - 1); i++) {
      const { r: r1, c: c1 } = amb.path[i];
      const { r: r2, c: c2 } = amb.path[i + 1];
      total += this.grid.getCongestion(r1, c1, r2, c2);
      count++;
    }
    return count > 0 && (total / count) > CONFIG.CONGESTION_THRESHOLD * 1.5;
  }
}
